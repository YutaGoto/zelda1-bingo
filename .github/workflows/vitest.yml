name: vitest

on:
  - push

jobs:
  vitest:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout your repository using git
        uses: actions/checkout@v4
      - name: Setup pnpm
        uses: pnpm/action-setup@v2
        with:
          version: 8
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          cache: "pnpm"
          node-version: 20
      - name: Install dependencies
        run: pnpm install
      - name: Run tests
        run: pnpm test
