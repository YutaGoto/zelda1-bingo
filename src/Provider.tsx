import {
  ChakraProvider,
  ColorModeScript,
  type ThemeConfig,
  extendTheme,
} from "@chakra-ui/react";
import { type StyleFunctionProps, mode } from "@chakra-ui/theme-tools";
import type { ReactNode } from "react";

interface ProviderProps {
  children: ReactNode;
}

export const Provider = ({ children }: ProviderProps) => {
  const config: ThemeConfig = {
    initialColorMode: "system",
    useSystemColorMode: true,
  };

  const theme = extendTheme({
    ...config,
    styles: {
      global: (props: StyleFunctionProps) => ({
        body: {
          bg: mode("gray.50", "gray.800")(props),
        },
      }),
    },
    fonts: {
      heading: `'Noto Sans JP Variable', sans-serif`,
      body: `'Noto Sans JP Variable', sans-serif`,
    },
  });

  return (
    <ChakraProvider theme={theme}>
      <ColorModeScript initialColorMode={config.initialColorMode} />
      {children}
    </ChakraProvider>
  );
};
