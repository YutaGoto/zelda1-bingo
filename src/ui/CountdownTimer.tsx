import { Box, Button, Group, Text } from "@chakra-ui/react";

import { convertToTime } from "../utils/convertToTime";
import { useCountdownTimer } from "../utils/useCountdownTimer";

export const CountdownTimer = () => {
  const { time, handleStart, handlePause, handleReset, isRunning } =
    useCountdownTimer({ initialTimeMinutes: 15 });

  return (
    <Box>
      <Box mb={1}>
        <Text fontSize="4xl" color={time === 0 ? "orange" : "default"}>
          {convertToTime({ time: time * 1000, isMillisecond: false })}
        </Text>
      </Box>

      <Group>
        <Button
          variant="outline"
          colorPalette="cyan"
          fontWeight={"bold"}
          onClick={() => handleStart()}
          disabled={isRunning}
        >
          Start
        </Button>
        <Button
          variant="outline"
          colorPalette="cyan"
          fontWeight={"bold"}
          onClick={() => handlePause()}
        >
          Pause
        </Button>
        <Button
          variant="outline"
          colorPalette="yellow"
          fontWeight={"bold"}
          onClick={() => handleReset()}
        >
          Reset
        </Button>
      </Group>
    </Box>
  );
};
