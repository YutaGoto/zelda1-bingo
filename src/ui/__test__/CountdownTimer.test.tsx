import { act, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { expect, test } from "vitest";

import { Provider } from "../../Provider";
import { CountdownTimer } from "../CountdownTimer";

const setup = () => {
  const user = userEvent.setup();
  return { user };
};

test("CountdownTimer", () => {
  render(<CountdownTimer />, { wrapper: Provider });
  expect(screen.getByText("Start")).toBeTruthy();
  expect(screen.getByText("Pause")).toBeTruthy();
  expect(screen.getByText("Reset")).toBeTruthy();
});

test("Snapshot", () => {
  const { asFragment } = render(<CountdownTimer />, { wrapper: Provider });
  expect(asFragment()).toMatchSnapshot();
});

test("Start", async () => {
  const { user } = setup();
  render(<CountdownTimer />, { wrapper: Provider });

  await act(async () => {
    await user.click(screen.getByText("Start"));
    await new Promise((r) => setTimeout(r, 2000));
  });

  expect(screen.queryByText("0:15:00")).toBeNull();
});
