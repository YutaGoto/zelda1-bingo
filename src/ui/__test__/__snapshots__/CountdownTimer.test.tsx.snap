// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;{let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&m?d.map(f=&gt;m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-0"
  >
    <div
      class="css-1wdlcwn"
    >
      <p
        class="css-1et0nzh"
      >
        0:15:00
      </p>
    </div>
    <div
      class="chakra-group css-6l0f26"
    >
      <button
        class="chakra-button css-1uutad9"
        data-first=""
        data-group-item=""
        style="--group-count: 3; --group-index: 0;"
        type="button"
      >
        Start
      </button>
      <button
        class="chakra-button css-1uutad9"
        data-between=""
        data-group-item=""
        style="--group-count: 3; --group-index: 1;"
        type="button"
      >
        Pause
      </button>
      <button
        class="chakra-button css-p45kro"
        data-group-item=""
        data-last=""
        style="--group-count: 3; --group-index: 2;"
        type="button"
      >
        Reset
      </button>
    </div>
  </div>
</DocumentFragment>
`;
